server/server_test.go

```diff
@@ -0,0 +1,109 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestNew(t *testing.T) {
+	cfg := Config{
+		HTTPListenAddr:                "localhost",
+		HTTPListenPort:                8080,
+		ServerGracefulShutdownTimeout: time.Second * 5,
+		HTTPServerReadTimeout:         time.Second * 10,
+		HTTPServerWriteTimeout:        time.Second * 10,
+		HTTPServerIdleTimeout:         time.Second * 15,
+	}
+
+	server, err := New(cfg)
+	if err != nil {
+		t.Fatalf("Failed to create server: %v", err)
+	}
+
+	if server.HTTPServer.Addr != fmt.Sprintf("%s:%d", cfg.HTTPListenAddr, cfg.HTTPListenPort) {
+		t.Errorf("Expected server address to be %s:%d, but got %s", cfg.HTTPListenAddr, cfg.HTTPListenPort, server.HTTPServer.Addr)
+	}
+}
+
+func TestRunAndShutdown(t *testing.T) {
+	cfg := Config{
+		HTTPListenAddr:                "localhost",
+		HTTPListenPort:                8081,
+		ServerGracefulShutdownTimeout: time.Second * 5,
+		HTTPServerReadTimeout:         time.Second * 10,
+		HTTPServerWriteTimeout:        time.Second * 10,
+		HTTPServerIdleTimeout:         time.Second * 15,
+	}
+
+	server, err := New(cfg)
+	if err != nil {
+		t.Fatalf("Failed to create server: %v", err)
+	}
+
+	go func() {
+		time.Sleep(time.Second)
```

Comment:
Using time.Sleep to synchronize goroutines is not a good idea. Better to use channels.

But again this server package looks very close to https://github.com/weaveworks/common/blob/master/server/server.go

I have a feeling we are going to use that in the end as we move forward. So let's not dwell too much on the sides for now

Reply:
Yeah, it looks very much like that file. The reason why I did not use that package is that it uses the gorilla/mux framework. Since I wanted to continue using the standard library, I wrote the server package